### ISSUE TITLE
C# Compiler shouldn't report CS8762 when calling function with NotNullWhen

### ISSUE THREAD

**Issue by ioklo** [dotnet/roslyn#45335]
**Version Used**: 3.1.**301**

**Steps to Reproduce**:
Write the code and compile
```
#nullable enable

class Program
{
    class C { }

    static bool InnerFunc([NotNullWhen(returnValue: true)] out C? outC)
    {
        outC = new C();
        return true;            
    }

    static bool Func1([NotNullWhen(returnValue: true)] out C? outC)
    {
        int i = 0;

        // Warning, CS8762
        return i switch
        {
            _ => InnerFunc(out outC)
        };
    }        

    static bool Func2([NotNullWhen(returnValue: true)] out C? outC)
    {
        var result = InnerFunc(out outC);

        // Warning, CS8762
        return result;
    }

    static bool Func3([NotNullWhen(returnValue: true)] out C? outC)
    {
        // No Warning
        return InnerFunc(out outC);
    }

    static void Main(string[] args)
    {
        Func1(out var c1);
        Func2(out var c2);
        Func3(out var c3);
    }
}
```

**Expected Behavior**: No Warnings, 3.1.**201** SDK doesn't report those warnings.

**Actual Behavior**: Reporting warnings in Func1 and Func2

**Comment by jaredpar** [comment]
@jcouv believe this is by design based and is just a result of the stricter method body checks that we do now. Can you confirm and close if that is the case?

**Comment by sharwell** [comment]
The warning in `Func2` is expected because the linked state associated with local variables doesn't propagate forward.

The warning in `Func1` seems like a possible oversight for switch expressions.

**Comment by jcouv** [comment]
I tested this just now ([sharplab](https://sharplab.io/#v2:CYLg1APgxAdgrgGwQQwEYIKYAIMzZgWAChiABABi1IEYA6AEQEtkBzGAewGcAXRgY060Awu2AYAgngQBPTo04BuYmQBMVagHZiAb2JZ9VNUKzasAX2VED6gGxZU7dgiwBJGDAwAnUgFYAFADaAHLs3EGICADqABa4fqQaAGrICHAYIFjcnmkAlAC6WOxw3FhCAPyFxUI5ega6VtYGRdzGALxYHgDupX45Sg2NVBqZ2RgKg4O1+hYkA7b2js4AYnAwfNSBIWERMXEJyanpI7kFzaUVzdVTJtfWjDAljFjt5P23BgD0H1iRyJ4w9xYABpSgBlAAcGhsKne+gSWCenE6jG4fGisJucwmWAA+s8AHyudxeXx+M6XGpYgxmfrWMzYyzWGh2BxOLArNYqTahcJIXYweJJFJpDJZE6VErlCVXOb1bEANz+WE8GE4iBK7TcHm8/nJVT6jImXx+fwBMGBYMh0Ix8JVaoQ3Fp1MNcOoLMW7NWfAAzNztnzYgL9sKjmKMPkJedpZTrHKjd8Qib/oCbcMtSTdcVo07pi75qQACxYACyyHu8Wo5ACBT+LE4MbqGI56zJWcVniw6wNVP0za5Z3bnZU3exzd9A6VPpHzqIZiAA=)) and see no warnings. 

We have relaxed the enforcement of NotNullWhen in method bodies recently (maybe 16.8?), so that we only warn when we are sure there's a problem. We only warn when we could determine that the nullability state of the WhenTrue and WhenFalse branches at the return statement are different. Typically, that means when returning a constant value or returning a method invocation that benefits from an attribute.

I'll go ahead and close the issue as fixed.
